aws/data_source_aws_regions_test.go

```diff
@@ -0,0 +1,119 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+)
+
+func TestAccDataSourceAwsRegions_Basic(t *testing.T) {
+	resourceName := "data.aws_regions.empty"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAwsRegionsConfig_empty(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceAwsRegions_OptIn(t *testing.T) {
+	resourceName := "data.aws_regions.opt_in_status"
+
+	statusOptedIn := "opted-in"
+	statusNotOptedIn := "not-opted-in"
+	statusOptInNotRequired := "opt-in-not-required"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			// This resource has to be at the very top of the test scenario due to bug in Terrafom Plugin SDK
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusOptedIn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusOptInNotRequired),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegionsFiltered(statusNotOptedIn),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceName),
+					resource.TestCheckNoResourceAttr(resourceName, "all_regions"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceAwsRegions_AllRegions(t *testing.T) {
+	resourceAllRegions := "data.aws_regions.all_regions"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceAwsRegionsConfig_allRegions(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccDataSourceAwsRegionsCheck(resourceAllRegions),
+					resource.TestCheckResourceAttr(resourceAllRegions, "all_regions", "true"),
+					resource.TestCheckNoResourceAttr(resourceAllRegions, "opt_in_status"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceAwsRegionsCheck(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("root module has no resource called %s", name)
+		}
+
+		return nil
+	}
+}
```

Comment:
This function should be augmented to check that results are returned as expected, e.g. that `names` exists and is greater than 0:

```suggestion
func testAccDataSourceAwsRegionsCheck(resourceName string) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		rs, ok := s.RootModule().Resources[resourceName]

		if !ok {
			return fmt.Errorf("root module has no resource called %s", resourceName)
		}
		
		names, namesOk := attrs["names.#"]

		if !namesOk {
			return fmt.Errorf("names attribute is missing.")
		}

		namesQuantity, err := strconv.Atoi(names)

		if err != nil {
			return fmt.Errorf("error parsing names (%s) into integer: %s", names, err)
		}

		if namesQuantity == 0 {
			return fmt.Errorf("No names found, this is probably a bug.")
		}

		return nil
	}
}
```